backend/src/main/java/com/example/backend/Specifications/FlightSpecifications.java

```diff
@@ -72,4 +75,97 @@ public static Specification<Flight> hasDepartureDate(LocalDate departureDate) {
     public static Specification<Flight> hasArrivalDateLessThan(LocalDate arrivalDate) {
         return (root, query, criteriaBuilder) -> criteriaBuilder.lessThan(root.get("arrivalDate"), arrivalDate);
     }
+
+    public static Specification<Flight> hasDepartureAirport(Integer id) {
+        return (root, query, criteriaBuilder) -> {
+            Join<Flight, FlightLeg> flightAndFlightLeg = root.join("flightLegs");
+            Join<FlightLeg, Airport> flightLegAndAirport = flightAndFlightLeg.join("departureAirport");
+            return criteriaBuilder.and(
+                    criteriaBuilder.equal(flightLegAndAirport.get("id"), id),
+                    criteriaBuilder.equal(flightAndFlightLeg.get("flightLegId"), 1));
+        };
+    }
+
+    public static Specification<Flight> hasArrivalAirport(Integer id) {
+        return (root, query, criteriaBuilder) -> {
+
+            if (query == null) {
+                throw new IllegalArgumentException("Query parameter cannot be null");
+            }
+
+            Subquery<Integer> subquery = createNumberOfLastFlightLegSubquery(root, query, criteriaBuilder);
+
+            Join<Flight, FlightLeg> flightAndFlightLeg = root.join("flightLegs");
+            Join<FlightLeg, Airport> flightLegAndAirport = flightAndFlightLeg.join("arrivalAirport");
+
+            return criteriaBuilder.and(
+                    criteriaBuilder.equal(flightAndFlightLeg.get("flightLegId"), subquery),
+                    criteriaBuilder.equal(flightLegAndAirport.get("id"), id));
+        };
+    }
+
+    public static Specification<Flight> hasAvailableSeats(SeatClass seatClass, int numberOfTickets) {
+        if (seatClass == SeatClass.ECONOMY)
+            return hasSeats("availableEconomySeats",numberOfTickets);
+        else
+            return hasSeats("availableBusinessSeats", numberOfTickets);
+    }
+
+    public static Specification<Flight> hasSeats(String attributeName, int numberOfTickets) {
+        return (root, query, criteriaBuilder) -> criteriaBuilder.greaterThanOrEqualTo(
+                root.get(attributeName),
+                numberOfTickets);
+    }
+
+    public static Specification<Flight> hasFlightType(FlightType flightType) {
+        if (flightType == FlightType.DIRECT)
+            return hasOneFlightLeg();
+        else
+            return hasMoreThanOneFlightLeg();
+    }
+
+    public static Specification<Flight> hasOneFlightLeg() {
+        return (root, query, criteriaBuilder) -> {
+            Subquery<Long> subquery = createLegCountSubquery(root, query, criteriaBuilder, true);
+            return criteriaBuilder.exists(subquery);
+        };
+    }
+
+    public static Specification<Flight> hasMoreThanOneFlightLeg() {
+        return (root, query, criteriaBuilder) -> {
+            Subquery<Long> subquery = createLegCountSubquery(root, query, criteriaBuilder, false);
+            return criteriaBuilder.exists(subquery);
+
+        };
+    }
+
```

Comment:
The same here, hasOneFlightLeg and hasMoreThanOneFlightLeg, can be one function.

Reply:
yes, you are right I fix it.